File: server/model/backend/git.js
const gitclient = require("nodegit"),
      toString = require('stream-to-string'),
      fs = require('fs'),
      Readable = require('stream').Readable,
      Path = require('path'),
      crypto = require("crypto"),
      BASE_PATH = "/tmp/";

let repos = {};
setInterval(() => autoVacuum, 1000*60*60*10);

module.exports = {
    test: function(params){
        if(!params || !params.repo){ return Promise.reject({message: 'invalid authentication', code: 'INVALID_PARAMS'}) };
        if(!params.commit) params.commit = "{action} ({filename}): {path}";
        if(!params.branch) params.branch = 'master';
        if(!params.author_name) params.author_name = "Nuage";
        if(!params.author_email) params.author_email = "https://nuage.kerjean.me";
        if(!params.committer_name) params.committer_name = "Nuage";
        if(!params.committer_email) params.committer_email = "https://nuage.kerjean.me";

        if(params.password && params.password.length > 2700){
            return Promise.reject({message: "Your password couldn\'t fit in a cookie :/", code: "COOKIE_ERROR"})
        }
        return git.open(params)
            .then(() => Promise.resolve(params));
    },
    cat: function(path, params){
        return git.open(params)
            .then((repo) => git.refresh(repo, params))
            .then(() => file.cat(calculate_path(params, path)));
    },
    ls: function(path, params){
        return git.open(params)
            .then((repo) => git.refresh(repo, params))
            .then(() => file.ls(calculate_path(params, path)))
            .then((files) => files.filter((file) => (file.name === '.git' && file.type === 'directory') ? false: true))
    },
    write: function(path, content, params){
        return git.open(params)
            .then(() => file.write(calculate_path(params, path), content))
            .then(() => git.save(params, path, "write"));
    },
    rm: function(path, params){
        return git.open(params)
            .then(() => file.rm(calculate_path(params, path)))
            .then(() => git.save(params, path, "delete"));
    },
    mv: function(from, to, params){
        return git.open(params)
            .then(() => file.mv(calculate_path(params, from), calculate_path(params, to)))
            .then(() => git.save(params, to, 'move'));
    },
    mkdir: function(path, params){
        return git.open(params)
            .then(() => file.mkdir(calculate_path(params, path)))
            .then(() => git.save(params, path, "create"))
    },
    touch: function(path, params){
        var stream = new Readable(); stream.push(''); stream.push(null);
        return git.open(params)
            .then(() => file.write(calculate_path(params, path), stream))
            .then(() => git.save(params, path, 'create'));
    }
};

function autoVacuum(){
    const MAXIMUM_DATE_BEFORE_CLEAN = new Date().getTime() - 1000*60*60*24;
    for(let repo_path in repos){
        if(repos[repo_path] > MAXIMUM_DATE_BEFORE_CLEAN){
            fs.unlink(repo_path);
            delete repos[repo_path];
        }
    }
}

function calculate_path(params, path){
    const repo = path_repo(params);
    const full_path = Path.posix.join(repo, path);
    if(full_path.indexOf(BASE_PATH) !== 0 || full_path === BASE_PATH){
        return BASE_PATH+"error";
    }
    return full_path;
}

function path_repo(obj){
    let hash = crypto.createHash('md5');
    for(let key in obj){
        if(typeof obj[key] === 'string'){
            hash.update(obj[key]);
        }
    }
    const path = BASE_PATH+"git_"+obj.uid+"_"+obj.repo.replace(/[^a-zA-Z]/g, "")+"_"+hash.digest('hex');
    repos[path] = new Date().getTime();
    return path;
}

const file = {};
file.write = function (path, stream){
    return new Promise((done, err) => {
        let writer = fs.createWriteStream(path, { flags : 'w' });
        stream.pipe(writer);
        writer.on('close', function(){
            done('ok');
        });
        writer.on('error', function(error){
            err(error);
        });
    });
};
file.mkdir = function(path){
    return new Promise((done, err) => {
        fs.mkdir(path, function(error){
            if(error){ return err(error); }
            return done("ok");
        });
    });
}
file.mv = function(from, to){
    return new Promise((done, err) => {
        fs.rename(from, to, function(error){
            if(error){ return err(error); }
            return done("ok");
        });
    });
}
file.ls = function(path){
    return new Promise((done, err) => {
        fs.readdir(path, (error, files) => {
            if(error){ return err(error); }
            Promise.all(files.map((file) => {
                return stats(path+file).then((stat) => {
                    stat.name = file;
                    return Promise.resolve(stat);
                });
            })).then((files) => {
                done(files.map((file) => {
                    return {
                        size: file.size,
                        time: new Date(file.mtime).getTime(),
                        name: file.name,
                        type: file.isFile()? 'file' : 'directory'
                    };
                }));
            }).catch((error) => err(error));
        });
    });

    function stats(path){
        return new Promise((done, err) => {
            fs.stat(path, function(error, res){
                if(error) return err(error);
                return done(res);
            });
        });
    }
}
file.rm = function(path){
    return rm(path);

    function rm(path){
        return stat(path).then((_stat) => {
            if(_stat.isDirectory()){
                return ls(path)
                    .then((files) => Promise.all(files.map(file => rm(path+file))))
                    .then(() => removeEmptyFolder(path));
            }else{
                return removeFileOrLink(path);
            }
        });
    }

    function removeEmptyFolder(path){
        return new Promise((done, err) => {
            fs.rmdir(path, function(error){
                if(error){ return err(error); }
                return done("ok");
            });
        });
    }
    function removeFileOrLink(path){
        return new Promise((done, err) => {
            fs.unlink(path, function(error){
                if(error){ return err(error); }
                return done("ok");
            });
        });
    }
    function ls(path){
        return new Promise((done, err) => {
            fs.readdir(path, function (error, files) {
                if(error) return err(error)
                return done(files)
            });
        });
    }
    function stat(path){
        return new Promise((done, err) => {
            fs.stat(path, function (error, _stat) {
                if(error){ return err(error); }
                return done(_stat);
            });
        });
    }
}

file.cat = function(path){
    return Promise.resolve(fs.createReadStream(path));
}


const git = {};
git.open = function(params){
    count = 0;
    return gitclient.Repository.open(path_repo(params))
        .catch((err) => {
            return gitclient.Clone(params.repo, path_repo(params), {fetchOpts: { callbacks: { credentials: git_creds.bind(null, params) }}})
                .then((repo) => {
                    const branch = params.branch;
                    return repo.getBranchCommit("origin/"+branch)
                        .catch(() => repo.getHeadCommit("origin"))
                        .then((commit) => {
                            return repo.createBranch(branch, commit)
                                .then(() => repo.checkoutBranch(branch))
                                .then(() => Promise.resolve(repo));
                        })
                        .catch(() => Promise.resolve(repo));
                });
        });
};

git.refresh = function(repo, params){
    count = 0;
    return repo.fetchAll({callbacks: { credentials: git_creds.bind(null, params) }})
        .then(() => repo.mergeBranches(params.branch, "origin/"+params.branch, gitclient.Signature.default(repo), 2))
        .catch(err => {
            if(err.errno === -13){
                return git.save(params, '', 'merge')
                    .then(() => git.refresh(repo, params))
                    .then(() => Promise.resolve(repo));
            }
            return Promise.resolve(repo);
        });
};

git.save = function(params, path = '', type = ''){
    count = 0;
    const author = gitclient.Signature.now(params.author_name, params.author_email);
    const committer = gitclient.Signature.now(params.committer_name, params.committer_email);
    const message = params.commit
          .replace("{action}", type)
          .replace("{dirname}", Path.posix.dirname(path))
          .replace("{filename}", Path.posix.basename(path))
          .replace("{path}", path || '');

    return git.open(params)
        .then((repo) => Promise.all([
            Promise.resolve(repo),
            getParent(repo, params),
            refresh(repo, params)
        ]))
        .then((data) => {
            const [repo, commit, oid] = data;
            const parents = commit ? [commit] : [];
            return repo.createCommit("HEAD", author, committer, message, oid, parents)
                .then(() => Promise.resolve(repo));
        })
        .then((repo) => {
            return repo.getRemote("origin")
                .then((remote) => {
                    return remote.push(
                        ["refs/heads/"+params.branch+":refs/heads/"+params.branch],
                        { callbacks: { credentials: git_creds.bind(null, params, true) }}
                    );
                })
                .catch((err) => Promise.reject({status: 403, message: "Not authorized to push"}));
        });

    function getParent(repo, params){
        return repo.getBranchCommit(params.branch)
            .catch(() => {
                return repo.getHeadCommit()
                    .catch(() => Promise.resolve(null));
            });
    }
    function refresh(repo, params){
        return repo.refreshIndex()
            .then((index) => {
                return index.addAll()
                    .then(() => index.write())
                    .then(() => index.writeTree());
            });
    }
};





// the count thinghy is used to see if the request succeeded or not
// when something fail, nodegit would just run the callback again and again.
// The only way to make it throw an error is to return the defaultNew thinghy
let count = 0;
function git_creds(params, fn, _count){
    count += 1;

    if(count > 1 && _count !== undefined){
        return new gitclient.Cred.defaultNew();
    }else if(/http[s]?\:\/\//.test(params.repo)){
        return new gitclient.Cred.userpassPlaintextNew(params.username, params.password);
    }else{
        return new gitclient.Cred.sshKeyMemoryNew(params.username, "", params.password, params.passphrase || "")
    }
}

File: server/model/backend/git.go
package backend

import (
	"fmt"
	. "github.com/mickael-kerjean/nuage/server/common"
	"golang.org/x/crypto/ssh"
	"gopkg.in/src-d/go-git.v4"
	"gopkg.in/src-d/go-git.v4/plumbing"
	"gopkg.in/src-d/go-git.v4/plumbing/object"
	"gopkg.in/src-d/go-git.v4/plumbing/transport"
	"gopkg.in/src-d/go-git.v4/plumbing/transport/http"
	sshgit "gopkg.in/src-d/go-git.v4/plumbing/transport/ssh"
	"io"
	"os"
	"path/filepath"
	"strings"
	"time"
)

const GitCachePath = "data/cache/git/"

var GitCache AppCache

type Git struct {
	git *GitLib
}

func init() {
	Backend.Register("git", Git{})

	GitCache = NewAppCache()
	cachePath := filepath.Join(GetCurrentDir(), GitCachePath)
	os.RemoveAll(cachePath)
	os.MkdirAll(cachePath, os.ModePerm)
	GitCache.OnEvict(func(key string, value interface{}) {
		g := value.(*Git)
		g.Close()
	})
}

type GitParams struct {
	repo           string
	username       string
	password       string
	passphrase     string
	commit         string
	branch         string
	authorName     string
	authorEmail    string
	committerName  string
	committerEmail string
	basePath       string
}

func (git Git) Init(params map[string]string, app *App) (IBackend, error) {
	if obj := GitCache.Get(params); obj != nil {
		return obj.(*Git), nil
	}
	g := &Git{
		git: &GitLib{
			params: &GitParams{
				params["repo"],
				params["username"],
				params["password"],
				params["passphrase"],
				params["commit"],
				params["branch"],
				params["authorName"],
				params["authorEmail"],
				params["committerName"],
				params["committerEmail"],
				"",
			},
		},
	}
	p := g.git.params
	if p.branch == "" {
		p.branch = "master"
	}
	if p.commit == "" {
		p.commit = "{action} ({filename}): {path}"
	}
	if p.authorName == "" {
		p.authorName = "Nuage"
	}
	if p.authorEmail == "" {
		p.authorEmail = "https://nuage.kerjean.me"
	}
	if p.committerName == "" {
		p.committerName = "Nuage"
	}
	if p.committerEmail == "" {
		p.committerEmail = "https://nuage.kerjean.me"
	}
	if len(params["password"]) > 2700 {
		return nil, NewError("Your password doesn't fit in a cookie :/", 500)
	}

	hash := GenerateID(params)
	p.basePath = GetAbsolutePath(GitCachePath + "repo_" + fmt.Sprint(hash) + "/")

	repo, err := g.git.open(p, p.basePath)
	g.git.repo = repo
	if err != nil {
		return g, err
	}
	GitCache.Set(params, g)
	return g, nil
}

func (g Git) Info() string {
	return "git"
}

func (g Git) Ls(path string) ([]os.FileInfo, error) {
	g.git.refresh()
	p, err := g.path(path)
	if err != nil {
		return nil, NewError(err.Error(), 403)
	}
	file, err := os.Open(p)
	if err != nil {
		return nil, err
	}
	return file.Readdir(0)
}

func (g Git) Cat(path string) (io.Reader, error) {
	p, err := g.path(path)
	if err != nil {
		return nil, NewError(err.Error(), 403)
	}
	return os.Open(p)
}

func (g Git) Mkdir(path string) error {
	p, err := g.path(path)
	if err != nil {
		return NewError(err.Error(), 403)
	}
	return os.Mkdir(p, os.ModePerm)
}

func (g Git) Rm(path string) error {
	p, err := g.path(path)
	if err != nil {
		return NewError(err.Error(), 403)
	}
	if err = os.RemoveAll(p); err != nil {
		return NewError(err.Error(), 403)
	}
	message := g.git.message("delete", path)
	if err = g.git.save(message); err != nil {
		return NewError(err.Error(), 403)
	}
	return nil
}

func (g Git) Mv(from string, to string) error {
	fpath, err := g.path(from)
	if err != nil {
		return NewError(err.Error(), 403)
	}
	tpath, err := g.path(to)
	if err != nil {
		return NewError(err.Error(), 403)
	}

	if err = os.Rename(fpath, tpath); err != nil {
		return NewError(err.Error(), 403)
	}
	message := g.git.message("move", from)
	if err = g.git.save(message); err != nil {
		return NewError(err.Error(), 403)
	}
	return nil
}

func (g Git) Touch(path string) error {
	p, err := g.path(path)
	if err != nil {
		return NewError(err.Error(), 403)
	}
	file, err := os.Create(p)
	if err != nil {
		return NewError(err.Error(), 403)
	}
	file.Close()

	message := g.git.message("create", path)
	if err = g.git.save(message); err != nil {
		return NewError(err.Error(), 403)
	}
	return nil
}

func (g Git) Save(path string, file io.Reader) error {
	p, err := g.path(path)
	if err != nil {
		return NewError(err.Error(), 403)
	}

	fo, err := os.Create(p)
	if err != nil {
		return err
	}
	io.Copy(fo, file)
	fo.Close()

	message := g.git.message("save", path)
	if err = g.git.save(message); err != nil {
		return NewError(err.Error(), 403)
	}
	return nil
}

func (g Git) Close() error {
	return os.RemoveAll(g.git.params.basePath)
}

func (g Git) path(path string) (string, error) {
	if path == "" {
		return "", NewError("No path available", 400)
	}
	basePath := filepath.Join(g.git.params.basePath, path)
	if string(path[len(path)-1]) == "/" {
		basePath += "/"
	}
	if strings.HasPrefix(basePath, g.git.params.basePath) == false {
		return "", NewError("There's nothing here", 403)
	}
	return basePath, nil
}

type GitLib struct {
	repo   *git.Repository
	params *GitParams
}

func (g *GitLib) open(params *GitParams, path string) (*git.Repository, error) {
	g.params = params

	if _, err := os.Stat(g.params.basePath); os.IsNotExist(err) {
		auth, err := g.auth()
		if err != nil {
			return nil, err
		}
		return git.PlainClone(path, false, &git.CloneOptions{
			URL:           g.params.repo,
			Depth:         1,
			ReferenceName: plumbing.ReferenceName(fmt.Sprintf("refs/heads/%s", g.params.branch)),
			SingleBranch:  true,
			Auth:          auth,
		})
	}
	return git.PlainOpen(g.params.basePath)
}

func (g *GitLib) save(message string) error {
	w, err := g.repo.Worktree()
	if err != nil {
		return NewError(err.Error(), 500)
	}
	_, err = w.Add(".")
	if err != nil {
		return NewError(err.Error(), 500)
	}

	_, err = w.Commit(message, &git.CommitOptions{
		All: true,
		Author: &object.Signature{
			Name:  g.params.authorName,
			Email: g.params.authorEmail,
			When:  time.Now(),
		},
		Committer: &object.Signature{
			Name:  g.params.committerName,
			Email: g.params.committerEmail,
			When:  time.Now(),
		},
	})
	if err != nil {
		return err
	}

	auth, err := g.auth()
	if err != nil {
		return err
	}
	return g.repo.Push(&git.PushOptions{
		Auth: auth,
	})
}

func (g *GitLib) refresh() error {
	w, err := g.repo.Worktree()
	if err != nil {
		return err
	}
	return w.Pull(&git.PullOptions{RemoteName: "origin"})
}

func (g *GitLib) auth() (transport.AuthMethod, error) {
	if strings.HasPrefix(g.params.repo, "http") {
		return &http.BasicAuth{
			Username: g.params.username,
			Password: g.params.password,
		}, nil
	}
	isPrivateKey := func(pass string) bool {
		if len(pass) > 1000 && strings.HasPrefix(pass, "-----") {
			return true
		}
		return false
	}

	if isPrivateKey(g.params.password) {
		signer, err := ssh.ParsePrivateKeyWithPassphrase([]byte(g.params.password), []byte(g.params.passphrase))
		if err != nil {
			return nil, err
		}
		return &sshgit.PublicKeys{
			User:   "git",
			Signer: signer,
			HostKeyCallbackHelper: sshgit.HostKeyCallbackHelper{
				HostKeyCallback: ssh.InsecureIgnoreHostKey(),
			},
		}, nil
	}

	return &sshgit.Password{
		User:     g.params.username,
		Password: g.params.password,
		HostKeyCallbackHelper: sshgit.HostKeyCallbackHelper{
			HostKeyCallback: ssh.InsecureIgnoreHostKey(),
		},
	}, nil
}

func (g *GitLib) message(action string, path string) string {
	message := strings.Replace(g.params.commit, "{action}", "save", -1)
	message = strings.Replace(message, "{filename}", filepath.Base(path), -1)
	message = strings.Replace(message, "{path}", strings.Replace(path, g.params.basePath, "", -1), -1)
	return message
}
